Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py

```diff
@@ -90,20 +91,64 @@ def editable(cls, branch, repository=None):
         return True
 
     @classmethod
-    def branch_point(cls, repository, limit=None):
-        cnt = 0
-        commit = repository.commit(include_log=False, include_identifier=False)
-        while cls.editable(commit.branch, repository=repository):
-            cnt += 1
-            if limit and cnt > limit:
-                return None
-            commit = repository.find(argument='HEAD~{}'.format(cnt), include_log=False, include_identifier=False)
-            if cnt > 1 or commit.branch != repository.branch or cls.editable(commit.branch, repository=repository):
-                log.info('    Found {}...'.format(string_utils.pluralize(cnt, 'commit')))
-            else:
-                log.info('    No commits on editable branch')
+    def branch_point(cls, repository):
+        branches = repository.branches_for(remote=None)
+        production_branches = [
+            'remotes/{}/{}'.format(remote, branch)
+            for remote in repository.source_remotes()
+            for branch in branches[remote] if not repository.dev_branches.match(branch)
+        ]
+
+        head = run(
+            [repository.executable(), 'rev-parse', 'HEAD'],
+            cwd=repository.root_path,
+            capture_output=True,
+            encoding='utf-8',
+        ).stdout.strip()
 
-        return commit
+        partial_bases = set()
+        for shard in [
+            production_branches[cls.MERGE_BASE_SHARD_SIZE * i:cls.MERGE_BASE_SHARD_SIZE * (i + 1)]
+            for i in range(1 + len(production_branches) // cls.MERGE_BASE_SHARD_SIZE)
+        ]:
+            if not shard:
+                continue
+            result = run(
+                [repository.executable(), 'merge-base', head] + shard,
+                cwd=repository.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            )
+            if result.returncode:
+                partial_bases = set()
+                break
+            partial_base = result.stdout.strip()
+            if partial_base == head:
+                return repository.commit(
+                    hash=head,
+                    include_log=False, include_identifier=False,
+                )
```

Comment:
A comment on why it's safe to prematurely return here may be helpful. IIUC, if, for instance, HEAD is the same commit as remotes/origin/main, then we already know that HEAD is the branch point and don't need to consider results from other shards.